Add default board to new projects, remove uncategorized pseudo-board

denyskon opened the pull request:
On creation of an empty project (no template) a default board will be created instead of falling back to the uneditable pseudo-board.

Every project now has to have exactly one default boards. As a consequence, you cannot unset a board as default, instead you have to set another board as default. Existing projects will be modified using a cron job, additionally this check will run every midnight by default.

Deleting the default board is not allowed, you have to set another board as default to do it.

Fixes #29873
Fixes #14679 along the way
Fixes #29853

Co-authored-by: delvh <dev.lh@web.de>

delvh commented:
Wait, the default board is a pseudo-board?
Where is it set?

denyskon commented:
@delvh https://github.com/go-gitea/gitea/blob/16e360099d0a515d429538ec88cff1f3ede23fb4/models/project/board.go#L258

lng2020 reviewed (changes requested):
I think it's not right.
The default board(which is called `uncategorized`) is not in the database. When the front end passes the default board id(which is 0), the backend always needs to check if the `default board id == 0` and if so, return a newly created default board. 
So if `case BoardTypeNone` is changed, the `default board id == 0` will never come true. Then all backend check logic should be changed/removed and then the DB table should be migrated.
This uneditable pseudo-board may not be a good design but it already exists. So we decided to refactor it, and the proper plan(or migration) should be designed.

denyskon commented:
@lng2020 The default logic will apply if all columns are deleted which in fact can easily happen. Also, we cannot possibly do this change retroactively for all existing projects which currently display this default pseudo-column.

lng2020 reviewed (commented):
> Also, we cannot possibly do this change retroactively for all existing projects which currently display this default pseudo-column.

You're right. Though I still doubt if this change breaks something because the `default project board id == 0` trick is used massively in codebase. I will dismiss my block and do some tests later.

lng2020 reviewed (commented):
Now I think the code logic is right. It creates a column with a given name and is set to default upon creating a project, with nothing breaking.
However, I found the default name `Uncategorized` a little confusing. Especially when users delete the deletable `Uncategorized` col, then another undeletable `Uncategorized` col appears. 
What about following the proposal issue #14679 that renaming it to `Backlog`? It seems reasonable to me that after deleting a default board, all issues of the default board become uncategorized.

lng2020 reviewed (approved):
The later [check](https://github.com/go-gitea/gitea/blob/16e360099d0a515d429538ec88cff1f3ede23fb4/models/project/board.go#L126-L128) seems redundant now. 
```
if len(items) == 0 {
	return nil
}
```
and would mind if you changed the title to `Allow setting default board for new projects`? I think it's clearer. What do you think?

lunny reviewed (changes requested):
Do we really need the configuration? Maybe we can hardcode it and make it an option when we think it's necessary.

denyskon commented:
@lunny You can already configure the default column names for the two project templates, I don't see a reason for the empty template column not to be configurable too.

denyskon commented:
I tried to write a migration however this is my first time doing such a complex one and something is wrong....

```go
func AddDefaultProjectBoard(x *xorm.Engine) error {
	sess := x.NewSession()
	defer sess.Close()

	if err := sess.Begin(); err != nil {
		return err
	}

	limit := setting.Database.IterateBufferSize
	if limit <= 0 {
		limit = 50
	}

	start := 0

	for {
		var projects []*project.Project
		if err := sess.Select("id, creator_id").Limit(limit, start).Find(&projects); err != nil {
			return err
		}

		if len(projects) == 0 {
			break
		}
		start += len(projects)

		for _, p := range projects {
			var board project.Board
			exist, err := sess.Where("project_id=? AND `default`=?", p.ID, true).Get(&board)
			if err != nil {
				return err
			}

			if !exist {
				board := project.Board{
					CreatedUnix: timeutil.TimeStampNow(),
					CreatorID:   p.CreatorID,
					Title:       setting.Project.ProjectBoardDefault,
					ProjectID:   p.ID,
					Default:     true,
				}
				if err := db.Insert(context.Background(), board); err != nil {
					return err
				}
			}
		}

		if start%1000 == 0 {
			if err := sess.Commit(); err != nil {
				return err
			}
			if err := sess.Begin(); err != nil {
				return err
			}
		}
	}

	return sess.Commit()
}
```

denyskon commented:
Okay, complete change of the purpose of this PR:

- removed the default ID 0 pseudo-column
- every project now needs to have a default column -> migration to add it
- forbid deleting default column
- remove unset default -> instead you need to set another column as default

denyskon commented:
@lunny Removed the config param

delvh commented:
Isn't this problem simply
```go
var projectsWithMissingDefault []int64
x.SQL(`SELECT DISTINCT p.project_id FROM project p WHERE NOT EXISTS (SELECT 1 FROM project_boards pb WHERE pb.project_id = p.project_id AND pb.default != 0);").Find(&projectsWithMissingDefault)
for _, projectID := range projectsWithMissingDefault {
  x.Insert(Board{ProjectID: projectID, Default: true, Title: …,  CreatorID: …})
}
```
?

denyskon commented:
@delvh The migration works in its current form, however that is a lot more lines of code than your solution :laughing: 

Wasn't it discouraged to write direct SQL in xorm queries? Or do I recall it wrongly?

Anyway, why `pb.default != 0`?

silverwind commented:
> Wasn't it discouraged to write direct SQL in xorm queries? Or do I recall it wrongly?

Depends if your SQL is portable. In my [last migration](https://github.com/go-gitea/gitea/pull/27419/files#diff-132e5de2b34cc8bc0f18346b2ae6868f54d8351d5470a12f7151e237a64c0cf6R10) it was so raw sql was fine there.

delvh commented:
> Wasn't it discouraged to write direct SQL in xorm queries?

Yep. However,
> Note that we might perhaps be able to beautify this snippet, although I don't know how

> Anyway, why pb.default != 0?

`pb.default` is an int, so `0` == `false`, and in other words `!= 0` == `true` as we are in a `NOT EXISTS`.

delvh commented:
> Depends if your SQL is portable

According to ChatGPT and my quick internet searches, it should be supported in MySQL, Postgres, and SQLite, so all our supported DBs.
However, quoting might be necessary.

silverwind commented:
> However, quoting might be necessary.

Quote field names in backtick and strings in single quote. I heard gorm takes care of specific quoting requirements of MSSQL and such when using that.

denyskon commented:
@delvh I was not able to make your solution work, at least not on sqlite. Could you try editing it yourself? Alternatively, we can keep my solution

delvh commented:
@denyskon I've adapted my code now, I can confirm it works now under SQLite.
My only worry is the escaping: I think some DBs use`"` instead of `` ` `` as escape mechanism.
Or does xorm convert `` ` `` to `"` automatically when required by the DB?
I'm never sure of that.

silverwind commented:
> @denyskon I've adapted my code now, I can confirm it works now under SQLite. My only worry is the escaping: I think some DBs use`"` instead of `` ` `` as escape mechanism. Or does xorm convert `` ` `` to `"` automatically when required by the DB? I'm never sure of that.

Pretty sure it does, I think it even uses `[value]` for MSSQL. @lunny knows for sure.

lunny commented:
Yes, ``` is a special quote that will be replaced by the database special quote mark automatically before the SQL will be executed. And if you know the database type, you can also use the original quote mark directly.

lunny commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,72 @@
+		}
+
+		if start%1000 == 0 {
+			if err := sess.Commit(); err != nil {
```
some database will be failed if we submit an empty transaction.

lunny commented:
It's better to have a test for `models/migrations/v1_22/v292.go`

denyskon commented:
@delvh Thanks, your migration works. Changed :)

denyskon commented:
@lunny I don't understand what you mean, the migration is now pretty simple, and I also have no idea how to test migrations....

silverwind commented on templates/projects/view.tmpl:
```diff
@@ -93,22 +93,15 @@
 									{{svg "octicon-trash"}}
 									{{ctx.Locale.Tr "repo.projects.column.delete"}}
 								</a>
+								{{end}}
```
Indent above, I think it's justified here because code above also indents.

delvh commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,34 @@
+
+func AddDefaultProjectColumn(x *xorm.Engine) error {
+	var projectsWithMissingDefault []project.Project
+	x.SQL("SELECT DISTINCT `p`.`id`, `p`.`creator_id` FROM `project` `p` WHERE NOT EXISTS (SELECT 1 FROM `project_board` `pb` WHERE `pb`.`project_id` = `p`.`id` AND `pb`.`default` != 0);").
```
I am wondering if the `DISTINCT` is still working correctly.
If any ID appears twice with different creator, we get an inconsistent state with 2 default boards.
On the other hand, I don't see how the creator can differ for the same project…

denyskon replied:
Consistency check added for multiple default boards

delvh reviewed (approved):
I strongly think we need a consistency check that there is **exactly** 1 default board per project. If no board is the default, one should be created. If more than 1 board is the default, the remaining ones need to be unset.

However, not sure if we need it in this PR.

wxiaoguang commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,34 @@
+
+func AddDefaultProjectColumn(x *xorm.Engine) error {
+	var projectsWithMissingDefault []project.Project
+	x.SQL("SELECT DISTINCT `p`.`id`, `p`.`creator_id` FROM `project` `p` WHERE NOT EXISTS (SELECT 1 FROM `project_board` `pb` WHERE `pb`.`project_id` = `p`.`id` AND `pb`.`default` != 0);").
```
`bool` is tricky in RDBMS, there is no standard support, so it can't use ` != 0`.

denyskon replied:
What should be used instead? `!= ?)", false` ?

wxiaoguang commented:
It seems that there is a hard requirement for "there must be one default board" ?

Could it relax the requirement and make the code more robust. For example, do not force users to set one default board, but only "select" a default board: use the one with `default=true` if exist, otherwise use the oldest board?

Then the code could be simplified a lot IMO.

wxiaoguang commented:
One more question, is it a breaking change? For example: if I was using "uncategorized" board, and added some issues into it. After this PR, where would these added uncategorized issues go? Could I still see them in a board?

silverwind commented:
> It seems that there is a hard requirement for "there must be one default board" ?
> 
> Could it relax the requirement and make the code more robust. For example, do not force users to set one default board, but only "select" a default board: use the one with `default=true` if exist, otherwise use the oldest board?
> 
> Then the code could be simplified a lot IMO.

I think it's totally fine if we would start new projects with empty boards.

lunny commented:
> One more question, is it a breaking change? For example: if I was using "uncategorized" board, and added some issues into it. After this PR, where would these added uncategorized issues go? Could I still see them in a board?

Yes, and users may also have a board named `Backlog`. So there will be two `Backlog` columns.

delvh reviewed (changes requested):
Apparently my comment above should not be postponed but done in this PR

denyskon commented:
@wxiaoguang
@lunny

Currently, the default "Uncategorized" board that you see when opening a project without a default board doesn't actually exist in DB, it is generated dynamically on render when opening a project which doesn't have a default board. Every issue not assigned to another board is automatically on "Uncategorized".

What this PR does is replace it with a default "Backlog" board for every project instead. This new board does exist in the DB, so it can be treated as every other board and e. g. be renamed, change colour and be deleted if you set any other board as default. Every issue that previously appeared on "Uncategorized" will now appear on "Backlog" -> for the users, the only noticeable changes will be the rename and the ability to rename the board, nothing else will change with the behaviour.

Two "Backlog" columns? It can be renamed then, you can also have two "Uncategorized" columns currently.

Use the oldest board as default if no default is set? I don't think it's something I would expect as a user. What is the problem with forcing a default board per project? It isn't impacting the user's options in any way. By enforcing it, we ensure that new issues appear where they are expected to instead of a random board which was never ever set as default.

I'm convinced this change is right in its current way. I'm open to other opinions, but please provide reasons why some of the things above are false assumptions :)

denyskon commented:
@delvh How would multiple defaults per project happen? It is enforced by our template that only one default board can be set.

And if multiple would exist, which would you want to be unset? The oldest ones? The least-oldest ones? Alphabetically? 😅

delvh commented:
Not sure, some sort of bug within our application, someone calling the wrong route manually without the help of the UI, someone manually tempering the DB, …
As you can see, I can think of a few reasons, although I agree that neither is likely to happen.

Regarding which one: I don't care, as far as I'm concerned, when this happens, you can choose one at random…
But if it does happen, it's probably best to only keep the board that was set as default first (we can probably only ensure it by checking when a project board has been updated. That may catch false positives, but I don't want to store extra timestamps only for this impossible error case).

lunny commented:
> Two "Backlog" columns? It can be renamed then, you can also have two "Uncategorized" columns currently.

Maybe use `Uncategorized` as the default column name is better than `Backlog` for existing boards. Some of them have columns named `Backlog` which will make them confusing. I guess nobody will name a created new column with `Uncategorized` because once he created a board, he will see a default column `Uncategorized`.

wxiaoguang commented:
> Every issue that previously appeared on "Uncategorized" will now appear on "Backlog" -> for the users, the only noticeable changes will be the rename and the ability to rename the board, nothing else will change with the behaviour.

Which code does so or is there any test code? I can only see the migration code adds a board, but I didn't see how the old issues were moved.


> Use the oldest board as default if no default is set? I don't think it's something I would expect as a user. What is the problem with forcing a default board per project? It isn't impacting the user's options in any way.

I meant we could make the code more robust. We still add a default "Uncategorized" board for legacy project. So ideally in most (all) cases there are still default boards. I just meant that we could add a fallback, to avoid unable to find a default board in daily usage.

denyskon commented:
> Which code does so or is there any test code? I can only see the migration code adds a board, but I didn't see how the old issues were moved.

As I said, the "Uncategorized" board doesn't exist - it's just a placeholder which contains all the issues that belong to a project, but don't belong to a board. At the moment when you set another board as default, the same behaviour applies to it too, so all the issues which appeared at "Uncategorized" should appear at "Backlog" without any moving needed :)

> I meant we could make the code more robust. We still add a default "Uncategorized" board for legacy project. So ideally in most (all) cases there are still default boards. I just meant that we could add a fallback, to avoid unable to find a default board in daily usage.

By our current code, there is no way it could happen, as the default board is created on project creation and the existing projects are updated using a migration.

denyskon commented:
> Which code does so or is there any test code? I can only see the migration code adds a board, but I didn't see how the old issues were moved.

As I said, the "Uncategorized" board doesn't exist - it's just a placeholder which contains all the issues that belong to a project, but don't belong to a board. At the moment when you set another board as default, the same behaviour applies to it too, so all the issues which appeared at "Uncategorized" should appear at "Backlog" without any moving needed :)

> I meant we could make the code more robust. We still add a default "Uncategorized" board for legacy project. So ideally in most (all) cases there are still default boards. I just meant that we could add a fallback, to avoid unable to find a default board in daily usage.

By our current code, there is no way it could happen, as the default board is created on project creation and the existing projects are updated using a migration.

wxiaoguang commented:
OK, I have no more questions (the SQL problem needs to be fixed, and I think it needs some tests)

denyskon commented:
> Maybe use Uncategorized as the default column name is better than Backlog for existing boards. Some of them have columns named Backlog which will make them confusing. I guess nobody will name a created new column with Uncategorized because once he created a board, he will see a default column Uncategorized.

Maybe use Uncategorized for existing projects, but Backlog for new projects? We had requests to rename the default board to backlog

denyskon commented:
@delvh Added a migration method to unset wrong default columns.

denyskon commented:
@wxiaoguang What would you like to be tested? The migration, the automatic creation of a default column....?

wxiaoguang commented:
> @wxiaoguang What would you like to be tested? The migration, the automatic creation of a default column....?

For example, the migration code? It looks quite complex.

And should it use `default =true` to replace `default != false` in case `default is null in database`? Not sure, just a question.

wxiaoguang commented on models/project/board.go:
```diff
@@ -255,16 +269,10 @@ func (p *Project) getDefaultBoard(ctx context.Context) (*Board, error) {
-		Title:     "Uncategorized",
-		Default:   true,
-	}, nil
+	return nil, fmt.Errorf("getDefaultBoard: no default board found")
```
That's what I worried about robust. If there is no default board (by any mistake), then the users would always see a 500 page, there is no way for them to recover.

denyskon replied:
I understand, but this case should ideally never happen - and if it does, it's a bug which we want for the users to report... 🤔

denyskon replied:
I think it is solved when such cases are checked every midnight by a cron job

delvh commented:
>  Added a migration method to unset wrong default columns.

Ah, I didn't mean a migration but a consistency check instead, so that it is periodically checked as a cronjob.

denyskon commented:
Okay so I did as suggested and changed the approach a bit. I added a cron job running every midnight, which ensures that every project has exactly one default column. As it also will be executed on initial start, there is no longer a need for a migration. I also added a test for the corresponding method. As this consistency check will run on every restart and also daily, I think it also solves the robustness problem if such one is ever to exist.

silverwind commented:
I welcome the removal of the migration, but does this mean a project still can not have no columns? Not a blocker for me but I guess there is no reason not to allow such empty projects.

delvh commented:
Exactly, there is a cronjob to ensure that if one happens somehow, it will be fixed automatically.

silverwind commented:
But why do we do that? Can't one have a empty board, maybe just temporarily?

silverwind commented:
FWIW, I tried deleting the last board on GitHub and it does not let me, but also does not give an error. So I guess GitHub also has problems with empty projects 😆.

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -1391,18 +1391,15 @@ projects.type.basic_kanban = "Basic Kanban"
-projects.column.unset_default_desc = "Unset this column as default"
 projects.column.delete = "Delete Column"
-projects.column.deletion_desc = "Deleting a project column moves all related issues to 'Uncategorized'. Continue?"
+projects.column.deletion_desc = "Deleting a project column moves all related issues to the default column. Continue?"
```
It needs a new key

denyskon replied:
Why? New translations for existing keys now invalidate the old ones, so it isn't necessary.

wxiaoguang replied:
NO. I have explained many times, it doesn't "invalidate", unless I remember wrong or Crowdin has changed behaviors.

Especially in your issue #27700, I have said clearly about this problem. https://github.com/go-gitea/gitea/issues/27700#issuecomment-1771892952

Let's see a real example: `pulls.push_rejected` has been changed from "Merge failed" to "Push failed" in #29537, but in other languages, they are still the old translation, eg: `pulls.push_rejected=合并失败 (merge failed)`

I won't strongly say objection for this, because I guess I don't want to be considered to be unfriendly during code review, and non-English translations don't bother me. I am suggesting just because my philosophy is "do the right thing".

denyskon replied:
Crowdin sync seems broken completely currently - deleted translation keys stay in place in the translations currently, and it needs to be fixed ASAP. But invalidation works, I've seen it multiple times

wxiaoguang replied:
How it works? Why `pulls.push_rejected` hasn't been updated in weeks .....

denyskon replied:
Currently, nothing works because (at least it seems to me) like our translations aren't synced to Crowdin at all. @techknowlogick should know more... we should switch to the Crowdin GitHub action ASAP

wxiaoguang replied:
Hmm, it seems that we have 2 problems now. 

1. The crowdin sync stops working
2. The translations don't get invalidated

I know the "don't get invalidated" because when I was writing the "backport-locales.go" (#23633), I found that many translations are outdated, for example: `%s` numbers don't match. It is a quite longstanding problem (and I have mentioned it many times 😢 )

denyskon replied:
There was a long time when the translations were not invalidated, but then techknowlogick changed something on the sync side and newly changed translations were invalidated, but not the old ones. That's another issue I would like to solve soon as part of #27700

wxiaoguang replied:
Well, I didn't know the "something" has been changed, nor whether it really works as expectation. And actually I would expect to "git blame" first to fix the old outdated keys, before correct the keys invalidation ......

denyskon replied:
It will happen soon, but outside of this PR obviously :)

wxiaoguang replied:
Hmm .... one more question: if there is no fix before 1.22, then non-English users would see incorrect sentences?

denyskon replied:
We need a fix before 1.22, otherwise there's A LOT of wrong and/or missing translation keys....

wxiaoguang reviewed (commented):
And TBH I still think a self-healing system is better than a cronjob IMO. The cronjob does nothing in 99.99...% time, I guess it shouldn't need to add more and more no-op cronjobs in the future for similar cases.

lunny reviewed (changes requested):
A migration is necessary to create default column named `Uncategory` for those exists boards with no default column in database.

denyskon commented:
Honestly I'm tired of rewriting the behaviour over and over again.

Why is a migration still necessary? The cron job runs on start and also adds a default "Uncategorized" board to projects without one, that's exactly what was done in the migration before.

Why is a cronjob wrong? I don't expect a `getDefaultBoard` method to create or modify columns...

Who am I supposed to listen to, @delvh who asked for a cronjob or @wxiaoguang who is against one or @lunny who requests a migration? Don't take it personal, I just don't understand at this point what I am supposed to do to get this PR done which started with 10 LOC 🥲

lunny commented:
> Honestly I'm tired of rewriting the behaviour over and over again.
> 
> Why is a migration still necessary? The cron job runs on start and also adds a default "Uncategorized" board to projects without one, that's exactly what was done in the migration before.
> 
> Why is a cronjob wrong? I don't expect a `getDefaultBoard` method to create or modify columns...
> 
> Who am I supposed to listen to, @delvh who asked for a cronjob or @wxiaoguang who is against one or @lunny who requests a migration? Don't take it personal, I just don't understand at this point what I am supposed to do to get this PR done which started with 10 LOC 🥲

Sorry for the confusion. Cronjob is not suitable for a migration task and we never do like that before.

We can create a migration and also provide a doctor command but not the cron job. I cannot imagine we will have more and more cron jobs to check the inconsistent problems in the background if we do that.

wxiaoguang commented:
> We can create a migration and also provide a doctor command but not the cron job.

The "doctor" system is also a disaster, unless it is really really necessary, I would suggest do not touch it too much. It is very difficult for end users to use, and not friendly in docker, not well tested and have many bugs.

That's why I suggest to make the system self-healing from beginning, just keep things simple. IMO a correct migration + self-healing + test-code is good enough.

delvh commented:
> IMO a correct migration + self-healing + test-code is good enough

Agreed.
I thought we could achieve this through a cronjob.
Also my bad, I was under the impression initially that the `doctor fix` commands were periodically executed in a cronjob which apparently isn't the case.

However, @wxiaoguang what exactly would be "self-healing" for you then, if not a cronjob?

silverwind commented:
Neither doctor nor cronjob are good solutions and I agree we should reduce those. A good place to fix would be a one-time action during startup, e.g. migration, or one-time action when a project is opened.

denyskon commented:
@lunny @wxiaoguang @delvh 

Under consideration of everything stated above:
1. Readded migration which adds the default column to projects without one and unsets columns as default if multiple are set
2. Added the same check to the getDefaultBoard method to run on every project view -> the error that there is no default column cannot appear any longer (is that the thing you called "self-healing"?)
3. Removed cron job
4. Added migration test and getDefaultBoard test

silverwind commented:
@lunny still blocking? Migration is back.

delvh commented on templates/projects/view.tmpl:
```diff
@@ -86,29 +86,22 @@
-									data-url="{{$.Link}}/{{.ID}}/unsetdefault">
-										{{svg "octicon-pin-slash"}}
-										{{ctx.Locale.Tr "repo.projects.column.unset_default"}}
+								{{end}}
+								{{if not .Default}}
```
```suggestion
```
It's the same `if`

lunny commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,57 @@
+			}
+		}
+
+		if _, err := x.Where(builder.Eq{"project_id": p.ID}.And(builder.In("id", boardsToUpdate))).
```
We should just create a default column for those board which has no default column. So why we should update these projects?

denyskon replied:
That was requested by @delvh as part of the consistency check

wxiaoguang commented on models/project/board.go:
```diff
@@ -244,27 +258,46 @@ func (p *Project) GetBoards(ctx context.Context) (BoardList, error) {
-	if exist {
+
+	// create a default board if none is found
+	if len(boards) == 0 {
```
As I said above, I think it is good to make the system self-healing. I strongly against to show a lot of unnecessary 500 errors to end users. Maybe TOC should have a final decision for this.

lunny replied:
But we don't know what happened. Maybe the last default record can be recovered? But we should not create a new one. I think it's good for self-healing. But we don't know whether creating a new one is right here.

wxiaoguang replied:
> But we don't know what happened.

Even if you know, what could you do?


> Maybe the last default record can be recovered? 

What's the difference? Even if a new board is created, the issues could also be moved to the correct one.

> Maybe the last default record can be recovered? 

I think it is good enough because I can't see a real bad case. Or it could add an error log (actually few people really look at server logs)

denyskon commented on routers/web/repo/projects.go:
```diff
@@ -628,28 +609,19 @@ func MoveIssues(ctx *context.Context) {
-			return
+	board, err := project_model.GetBoard(ctx, ctx.ParamsInt64(":boardID"))
+	if err != nil {
+		if project_model.IsErrProjectBoardNotExist(err) {
+			ctx.NotFound("ProjectBoardNotExist", nil)
+		} else {
+			ctx.ServerError("GetProjectBoard", err)
```
This file only uses this if-else way of constructing such errors. I don't want to add more unrelated changes, will refactor it in a batch afterward in another PR.

denyskon commented on routers/web/org/projects.go:
```diff
@@ -674,28 +655,19 @@ func MoveIssues(ctx *context.Context) {
-			return
+	board, err := project_model.GetBoard(ctx, ctx.ParamsInt64(":boardID"))
+	if err != nil {
+		if project_model.IsErrProjectBoardNotExist(err) {
+			ctx.NotFound("ProjectBoardNotExist", nil)
+		} else {
+			ctx.ServerError("GetProjectBoard", err)
```
As above

denyskon commented on models/project/board.go:
```diff
@@ -274,10 +307,8 @@ func SetDefaultBoard(ctx context.Context, projectID, boardID int64) error {
-			Cols("`default`").Update(&Board{Default: true})
-	}
+	_, err = db.GetEngine(ctx).ID(boardID).Where(builder.Eq{"project_id": projectID}).
+		Cols("`default`").Update(&Board{Default: true})
```
Added check in https://github.com/go-gitea/gitea/pull/29874/commits/5405384be360f28c3c7ef8ad986f029f98e94fcc

denyskon commented:
@go-gitea/technical-oversight-committee I would like to ask for a TOC vote on which of the two strategies should be continued with, as the discussion here doesn't seem to be going anywhere.

The two possible solutions are:
1. Keep the consistency check ("self-healing") on every project viewing -> don't display errors but try to fix them (suggested by @delvh and @wxiaoguang) -> current state of this PR
2. Only keep the migration, consider every inconsistency in DB an issue which should be reported instead of trying to fix it automatically (proposed by @lunny) -> basically a revert to the state of https://github.com/go-gitea/gitea/pull/29874#issuecomment-2016526148

If there is no decision made by the TOC I would like for the PR to be either merged as-it-is (it has two necessary approvals and no blockers) or to be abandoned if no consensus is found, however then we'll remain with a state nobody likes.

delvh commented on models/project/board.go:
```diff
@@ -307,6 +307,11 @@ func SetDefaultBoard(ctx context.Context, projectID, boardID int64) error {
+	_, err = GetBoard(ctx, boardID)
+	if err != nil {
+		return err
+	}
```
That call should be before `#302`, otherwise we have a project without default board again.

wxiaoguang replied:
Yup, that's what I mentioned before: there might be various cases that there is no default board (either by existing bug, or newly introduced bug in the future). So self-healing is good enough.

---

Or ideally they could be in a transaction, then no worry about inconsistency.

denyskon replied:
Moved

delvh reviewed (approved):
I'm fine either way.
This PR was already changed way too often.
So, from my side, let's merge it (after the small bugfix below).
I don't know what others think about it.

silverwind commented:
CI shows some data race error, not sure if related.

wxiaoguang commented:
> Unit tests show [some data race error](https://github.com/go-gitea/gitea/actions/runs/8426585553/job/23075228826?pr=29874), not sure if related. Restarted it once.

Not related

lunny commented:
> @go-gitea/technical-oversight-committee I would like to ask for a TOC vote on which of the two strategies should be continued with, as the discussion here doesn't seem to be going anywhere.
> 
> The two possible solutions are:
> 
>     1. Keep the consistency check ("self-healing") on every project viewing -> don't display errors but try to fix them (suggested by @delvh and @wxiaoguang) -> current state of this PR
> 
>     2. Only keep the migration, consider every inconsistency in DB an issue which should be reported instead of trying to fix it automatically (proposed by @lunny) -> basically a revert to the state of [Add default board to new projects, remove uncategorized pseudo-board #29874 (comment)](https://github.com/go-gitea/gitea/pull/29874#issuecomment-2016526148)
> 
> 
> If there is no decision made by the TOC I would like for the PR to be either merged as-it-is (it has two necessary approvals and no blockers) or to be abandoned if no consensus is found, however then we'll remain with a state nobody likes.

I will not block this.

6543 commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,58 @@
+// CheckProjectColumnsConsistency ensures there is exactly one default board per project present
+func CheckProjectColumnsConsistency(x *xorm.Engine) error {
+	var projects []project.Project
+	if err := x.SQL("SELECT DISTINCT `p`.`id`, `p`.`creator_id` FROM `project` `p` WHERE (SELECT COUNT(*) FROM `project_board` `pb` WHERE `pb`.`project_id` = `p`.`id` AND `pb`.`default` = ?) != 1", true).
```
I'd like to have pagination in this migration!

denyskon replied:
Honestly I have no idea what you mean or how to do it 🙈

6543 replied:
Itterate over it in batches ... just look at older migrations

denyskon replied:
Oh so the thing I did before 😆

6543 replied:
Uh you did?

delvh replied:
Yep, and then it was removed again.
Also, I don't quite see the benefit of batching here.
How does it help with anything?

silverwind replied:
I guess the only reason to batch is when amount of project boards is really massive and the SQL operation would risk running into timeouts maybe.

denyskon replied:
Done

silverwind commented:
@6543 looking good to you?

denyskon commented:
Finally 🎉

wxiaoguang commented on models/migrations/v1_22/v292.go:
```diff
@@ -0,0 +1,85 @@
+			Limit(limit, start).
+			Find(&projects); err != nil {
+			return err
+		}
```
Sorry but this migration seems broken ....

If the "SELECT" results are stable and never change, this `for` loop is right.

BUT, by each `SELECT`, the "project boards" are changed. The next `SELECT`, it should still process the first page results, but not skip `start`.

lunny replied:
I will send a PR to fix it.

lunny replied:
#30153

melroy89 replied:
I believe I still have an issue with this migration. It says: "Ensure every project has exactly one default column". It tries to do some SQL queriy on the `project` and `project_board` tables. Migration seems to fail with error: `pq: invalid input syntax for type bigint: "true"`. Help?? I'm stuck on migration. My docker container keeps restating in a loop now.

I'm using Docker `gitea/gitea:latest-rootless` image with PostgreSQL v14.

```
2024/06/12 13:41:30 ...ations/migrations.go:642:Migrate() [I] [SQL] SELECT tablename FROM pg_tables WHERE schemaname = $1 [public] - 371.738µs
2024/06/12 13:41:30 .../xorm@v1.3.8/sync.go:30:Sync() [I] [SQL] SELECT column_name, column_default, is_nullable, data_type, character_maximum_length, description,
    CASE WHEN p.contype = 'p' THEN true ELSE false END AS primarykey,
    CASE WHEN p.contype = 'u' THEN true ELSE false END AS uniquekey
FROM pg_attribute f
    JOIN pg_class c ON c.oid = f.attrelid JOIN pg_type t ON t.oid = f.atttypid
    LEFT JOIN pg_attrdef d ON d.adrelid = c.oid AND d.adnum = f.attnum
    LEFT JOIN pg_description de ON f.attrelid=de.objoid AND f.attnum=de.objsubid
    LEFT JOIN pg_namespace n ON n.oid = c.relnamespace
    LEFT JOIN pg_constraint p ON p.conrelid = c.oid AND f.attnum = ANY (p.conkey)
    LEFT JOIN pg_class AS g ON p.confrelid = g.oid
    LEFT JOIN INFORMATION_SCHEMA.COLUMNS s ON s.column_name=f.attname AND c.relname=s.table_name
WHERE n.nspname= s.table_schema AND c.relkind = 'r' AND c.relname = $1 AND s.table_schema = $2 AND f.attnum > 0 ORDER BY f.attnum; [version public] - 2.886482ms
2024/06/12 13:41:30 .../xorm@v1.3.8/sync.go:30:Sync() [I] [SQL] SELECT indexname, indexdef FROM pg_indexes WHERE tablename=$1 AND schemaname=$2 [version public] - 431.557µs
2024/06/12 13:41:30 routers/common/db.go:31:InitDBEngine() [W] Table version column id db type is BIGINT, struct type is BIGSERIAL
2024/06/12 13:41:30 ...orm@v1.3.8/engine.go:1252:Get() [I] [SQL] SELECT "id", "version" FROM "version" WHERE "id"=$1 LIMIT 1 [1] - 223.978µs
2024/06/12 13:41:30 ...ations/migrations.go:688:Migrate() [I] Migration[293]: Ensure every project has exactly one default column
2024/06/12 13:41:30 routers/common/db.go:46:migrateWithSetting() [I] [SQL] BEGIN TRANSACTION [] - 53.6µs
2024/06/12 13:41:30 ...ations/v1_22/v293.go:55:CheckProjectColumnsConsistency() [I] [SQL] SELECT project.id as id, project.creator_id, project_board.id as board_id FROM "project" LEFT JOIN "project_board" ON project_board.project_id = project.id AND project_board."default"=$1 WHERE (project_board.id is NULL OR project_board.id = 0) LIMIT 50 [true] - 210.268µs
2024/06/12 13:41:30 ...ations/migrations.go:691:Migrate() [I] [SQL] ROLLBACK [] - 54.09µs
2024/06/12 13:41:30 routers/common/db.go:36:InitDBEngine() [E] ORM engine initialization attempt #8/10 failed. Error: migrate: migration[293]: Ensure every project has exactly one default column failed: pq: invalid input syntax for type bigint: "true"
2024/06/12 13:41:30 routers/common/db.go:37:InitDBEngine() [I] Backing off for 3 seconds
2024/06/12 13:41:31 ...eful/manager_unix.go:144:handleSignals() [W] PID 7. Received SIGTERM. Shutting down...
```
